README.md
# cms-sq-sap-hana-rr-cluster_monitoring

CHANGELOG.md
# 0.1.0

Initial release of cms_sq_sap_hana_rr_cluster_monitoring

* Enhancements
  * an enhancement

* Bug Fixes
  * a bug fix


spec/spec_helper.rb
#
# Cookbook:: cms_sq_sap_hana_rr_cluster_monitoring
# Spec:: default
#
# Copyright 2018 IBM Corporation, All Rights Reserved

require 'chefspec'
require 'chefspec/berkshelf'

# The platforms we support
# For a complete list of available platforms and versions see:
# https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
def test_platforms
  {
    'redhat' => %w(6.8 7.3),
  }.each do |platform, versions|
    versions.each do |version|
      yield(platform, version)
    end
  end
end

ChefSpec::Coverage.start! do
  add_output do |report_output|
    File.open('coverage.json', 'w') do |f|
      f.puts(report_output[:coverage])
    end
  end
end


spec/unit/recipes/default_spec.rb
#
## Cookbook:: cms_sq_sap_hana_rr_cluster_monitoring
# Spec:: default
#
# Copyright 2018 IBM Corporation, All Rights Reserved

require 'spec_helper'

describe 'cms_sq_sap_hana_rr_cluster_monitoring::default' do
  context 'SUSE Linux Enterprise Server 12 (x86_64)' do
    context 'When the node is HAcluster' do
      let(:chef_run) do
        runner = ChefSpec::ServerRunner.new(platform: 'suse', version: '12.2') do |node|
          node.normal['cluster_readiness']['HAcluster'] = 'true'
          node.normal['authentication']['dc_hostname'] = 'dal09ammadc001.example.com'
          node.normal['authentication']['ads_org_unit'] = 'Infrastructure Servers'
          node.normal['chef_client']['config']['chef_server_url'] = 'https://chef.example.com/organizations/test'
          node.normal['authentication']['default_realm'] = 'IMZCLOUD.IBMAMMSAP.LOCAL'
        end
        runner.converge('fms-zabbix::default', described_recipe)
      end
      it 'creates the resource script file' do
        expect(chef_run).to create_cookbook_file('/var/lib/zabbix/resource_script.sh').with(
          owner: 'root',
          group: 'root',
          mode: '0755'
        )
      end
      it 'creates the log script file' do
        expect(chef_run).to create_cookbook_file('/var/lib/zabbix/log_script.sh').with(
          owner: 'root',
          group: 'root',
          mode: '0755'
        )
      end
      it 'creates the pacemaker template ' do
        expect(chef_run).to create_template('/etc/zabbix/zabbix.conf.d/pacemaker.conf').with(
          source: 'pacemaker.conf.erb',
          owner: 'root',
          group: 'root',
          mode: '0644'
        )
      end
    end
    context 'When the node is not HAcluster' do
      let(:chef_run) do
        runner = ChefSpec::ServerRunner.new(platform: 'suse', version: '12.2') do |node|
          node.normal['cluster_readiness']['HAcluster'] = 'false'
          node.normal['authentication']['dc_hostname'] = 'dal09ammadc001.example.com'
          node.normal['authentication']['ads_org_unit'] = 'Infrastructure Servers'
          node.normal['chef_client']['config']['chef_server_url'] = 'https://chef.example.com/organizations/test'
          node.normal['authentication']['default_realm'] = 'IMZCLOUD.IBMAMMSAP.LOCAL'
        end
        runner.converge('fms-zabbix::default', described_recipe)
      end
      it 'deletes the conf file' do
        expect(chef_run).to delete_file('/etc/zabbix/zabbix.conf.d/pacemaker.conf')
      end
      it 'deletes the resource script file' do
        expect(chef_run).to delete_file('/var/lib/zabbix/resource_script.sh')
      end
      it 'deletes the log script file' do
        expect(chef_run).to delete_file('/var/lib/zabbix/log_script.sh')
      end
    end
  end
end


test/integration/default/default_spec.rb
is_cluster = attribute('HAcluster', default: 'false', description: 'Running on cluster')
if is_cluster == 'true'
  describe file('/etc/zabbix/zabbix.conf.d/pacemaker.conf') do
    it { should exist }
    it { should be_file }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0644 }
  end
  describe file('/var/lib/zabbix/resource_script.sh') do
    it { should exist }
    it { should be_file }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0755 }
  end
  describe file('/var/lib/zabbix/log_script.sh') do
    it { should exist }
    it { should be_file }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0755 }
  end
else
  describe file('/etc/zabbix/zabbix.conf.d/pacemaker.conf') do
    it { should_not exist }
  end
  describe file('/var/lib/zabbix/resource_script.sh') do
    it { should_not exist }
  end
  describe file('/var/lib/zabbix/log_script.sh') do
    it { should_not exist }
  end
end


metadata.rb
name 'cms_sq_sap_hana_rr_cluster_monitoring'
maintainer 'Maintainer name here'
maintainer_email 'Your email here'
license 'Proprietary - All Rights Reserved'
description 'Installs/Configures cms_sq_sap_hana_rr_cluster_monitoring'
long_description 'Installs/Configures cms_sq_sap_hana_rr_cluster_monitoring'
version '0.1.0'
chef_version '>= 12.1' if respond_to?(:chef_version)

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
issues_url 'https://github.ibm.com/cms-infra-cookbooks/cms_sq_sap_hana_rr_cluster_monitoring/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
source_url 'https://github.ibm.com/cms-infra-cookbooks/cms_sq_sap_hana_rr_cluster_monitoring'

supports 'redhat', '>= 6.5'

depends 'fms-zabbix', '~> 0.3.0'


files/default/log_script.sh

#!/bin/bash

p_script="
infile = '/var/log/cluster/corosync.log'
import re
from datetime import datetime,timedelta
strptime = datetime.strptime
year=datetime.now().year
count=0
month=datetime.now().strftime('%b')
for line in reversed(open(infile).readlines()):
        if (line.startswith(str(month)) ):
                d1=' '.join((line[0:6],str(year)))
                d2=' '.join((d1,line[7:15]))
                sc_date = datetime.strptime(d2, '%b %d %Y %H:%M:%S')
                dif_date=datetime.now() - timedelta(minutes=5)
                if(sc_date >= dif_date ):
                        m = re.search(r'(crmd.*(warning|error))|(stonith.*(warning|error))|(pengine.*(warning|error))|(sbd.*(warning|error))|(cib.*(info|warning|error)) ',line,re.M|re.I)
                        if m:
                                print (line);
	
        else:
                continue
"
 p_out=$(python -c "$p_script")
    if [ $? -eq 0 ]
    then
        echo "$p_out"
    else
        echo "UNKNOWN: python returncode $?"
        exit ${STATE_UNKNOWN}
    fi


files/default/resource_script.sh
#!/bin/bash

while getopts “i:” OPTION; do
  case $OPTION in
    i)
      item=$OPTARG
      ;;
    ?)
      echo "UNKNOWN: Unknown argument: $1"
      exit $ST_UK
      ;;
  esac
done

# MANDATORY args
if [ -z "$item" ]; then
  echo "UNKNOWN: missing argument -i"
  exit $ST_UK
fi
# check for info and set $STATUS
out=$(crm_mon --one-shot --as-xml 2>/dev/null)
STATUS=$?

p_import="import xml.etree.cElementTree as et,sys,io;tree=et.ElementTree();"
p_tree="tree.parse(io.BytesIO(b'''$out'''));root=tree.getroot();"
if [ ${STATUS} -eq 0 ]
then
node_script="resource_list={};
node_list={};
pnode=dnode=unode='';
for value in root.findall('node_attributes/node'):
        node_list=dict([(j.get('name'), j.get('value')) for j in value]);
        resource_list[value.attrib['name']]=node_list;
        if ( node_list['hana_tha_clone_state'] == 'PROMOTED' ):
                pnode = value.attrib['name'];
        if (  node_list['hana_tha_clone_state'] == 'DEMOTED' ):
                dnode = value.attrib['name'];
        if (  node_list['hana_tha_clone_state'] == 'UNDEFINED' ):
                unode = value.attrib['name'];

"
case "$item" in
data_list)
p_script="
print reource_list
"
;;
nodes_online)
p_script="
print ' '.join([f.get('online') for f in root.findall('nodes/node')])"
;;
resources_running)
p_script="
list = ['Master','Slave','Started'];
count = 0;
for elem in tree.iter(tag='resource'):
        for i in list:
			 if ( elem.get('role') == i ):
                          count += 1
print count;
"
;;
p_node)p_script="
print pnode;"
;;
d_node)
p_script="
if ( len(resource_list) == 2 ):
        if dnode:
                print dnode;
        else:
                print unode;
else:
        print 'Node offline';
"
;;
p_role)
p_script="
if pnode:
        print resource_list[pnode]['hana_tha_roles'];
else:
        print resource_list[unode]['hana_tha_roles'];
"
;;
p_sync_state)
p_script="
if pnode:
        print resource_list[pnode]['hana_tha_sync_state'];
else:
        print resource_list[unode]['hana_tha_sync_state'];
"
;;
p_score)
p_script="
if pnode:
        print resource_list[pnode]['master-rsc_SAPHana_THA_HDB00'];
else:
        print resource_list[unode]['master-rsc_SAPHana_THA_HDB00'];
"
;;
d_role)
p_script="
if ( len(resource_list) == 2 ):
        if dnode:
                print resource_list[dnode]['hana_tha_roles'];
        else:
                print resource_list[unode]['hana_tha_roles'];

else:
        print 'Node offline';
"
;;
d_sync_state)
p_script="
if ( len(resource_list) == 2 ):
        if dnode:
                print resource_list[dnode]['hana_tha_sync_state'];
        else:
                print resource_list[unode]['hana_tha_sync_state'];
else:
        print 'Node offline';
"
;;
d_score)
p_script="
if ( len(resource_list) == 2 ):
        if dnode:
                print resource_list[dnode]['master-rsc_SAPHana_THA_HDB00'];
        else:
                print resource_list[unode]['master-rsc_SAPHana_THA_HDB00'];

else:
         print 'Node offline';
"
;;
*) echo "UNKNOWN: wrong item: $item"
 exit ${STATE_UNKNOWN}
;;
esac
p_out=$(python -c "$p_import $p_tree $node_script $p_script") # 2>/dev/null)
    if [ $? -eq 0 ]
    then
        echo "$p_out"
    else
        echo "UNKNOWN: python returncode $?"
        exit ${STATE_UNKNOWN}
    fi
elif [ ${STATUS} -eq 107 ]
then
    echo "Connection to cluster failed: returncode ${STATUS}"
else
    echo "UNKNOWN: returncode ${STATUS}"
 exit ${STATE_UNKNOWN}
fi

recipes/default.rb
#
# Cookbook Name:: cms_sq_sap_hana_rr_cluster_monitoring
# Recipe:: default
#
# Copyright 2018 IBM Corporation, All Rights Reserved

# TODO: write the cookbook
config_file = '/etc/zabbix/zabbix.conf.d/pacemaker.conf'
['resource_script.sh', 'log_script.sh'].each do |conf_file|
  cookbook_file "/var/lib/zabbix/#{conf_file}" do
    owner 'root'
    group 'root'
    mode '0755'
    action :create
    only_if { node['cluster_readiness']['HAcluster'] == 'true' }
  end
end
template config_file do
  source 'pacemaker.conf.erb'
  owner  'root'
  group  'root'
  mode   '0644'
  only_if { node['cluster_readiness']['HAcluster'] == 'true' }
  notifies :restart, 'service[zabbix-agent]', :delayed
end
[config_file, '/var/lib/zabbix/resource_script.sh', '/var/lib/zabbix/log_script.sh'].each do |conf_file|
  file conf_file do
    action :delete
    only_if { node['cluster_readiness']['HAcluster'] == 'false' }
    notifies :restart, 'service[zabbix-agent]', :delayed
  end
end
